clang/lib/AST/PreorderAST.cpp

```diff
@@ -139,40 +157,210 @@ void PreorderAST::Create(Expr *E, OperatorNode *Parent) {
+      }
+    }
+
+    if (ArrowBase) {
+      // If ArrowBase exists, then E is of the form ArrowBase->f,
+      // (*ArrowBase).f, etc. The Base of the MemberNode is ArrowBase + 0
+      // so that expressions such as a->f, (*a).f, (a + 0)->f, and a[0].f
+      // all have the same canonical form.
+      auto *N = new MemberNode(Field, /*IsArrow*/ true, Parent);
+      AddNode(N, Parent);
+      AddZero(ArrowBase, /*Parent*/ N);
+    } else {
+      // If no ArrowBase exists, then E is of the form a.f.
+      auto *N = new MemberNode(Field, /*IsArrow*/ false, Parent);
+      AddNode(N, Parent);
+      Create(Base, /*Parent*/ N);
+    }
+
+  } else if (auto *UO = dyn_cast<UnaryOperator>(E)) {
+    UnaryOperatorKind Op = UO->getOpcode();
+    if (Op == UnaryOperatorKind::UO_Deref) {
+      // The child of a dereference operator must be a binary operator so that
+      // *e and *(e + 0) have the same canonical form. So for an expression of
+      // the form *e, we create a UnaryOperatorNode whose child is a
+      // BinaryOperatorNode e + 0.
+      auto *Child = UO->getSubExpr();
+      auto *N = new UnaryOperatorNode(Op, Parent);
+      AddNode(N, Parent);
+      AddZero(Child, /*Parent*/ N);
+    } else if ((Op == UnaryOperatorKind::UO_Plus ||
+               Op == UnaryOperatorKind::UO_Minus) &&
+               E->isIntegerConstantExpr(Ctx)) {
+      // For integer constant expressions of the form +e or -e, we create a
+      // LeafExprNode rather than a UnaryOperatorNode so that these expressions
+      // can be constant folded. Constant folding only folds LeafExprNodes that
+      // are children of a BinaryOperatorNode.
+      auto *N = new LeafExprNode(E, Parent);
+      AddNode(N, Parent);
+    } else {
+      auto *N = new UnaryOperatorNode(Op, Parent);
+      AddNode(N, Parent);
+      Create(UO->getSubExpr(), /*Parent*/ N);
+    }
+
+  } else if (auto *AE = dyn_cast<ArraySubscriptExpr>(E)) {
+    // e1[e2] has the same canonical form as *(e1 + e2).
+    auto *DerefExpr = BinaryOperator::Create(Ctx, AE->getBase(), AE->getIdx(),
+                                             BinaryOperatorKind::BO_Add, AE->getType(),
+                                             AE->getValueKind(), AE->getObjectKind(),
+                                             AE->getExprLoc(), FPOptionsOverride());
+    auto *N = new UnaryOperatorNode(UnaryOperatorKind::UO_Deref, Parent);
+    AddNode(N, Parent);
+    // Even though e1 + e2 is already a binary operator, the child of the
+    // UnaryOperatorNode should be e1 + e2 + 0. This enables expressions such
+    // as p[i + -(1 + 2)] to be constant folded. In order for a
+    // BinaryOperatorNode to be constant folded, it must have at least two
+    // LeafExprNode children whose expressions are integer constants. For
+    // example, i + -(1 + 2) + 0 will be constant folded to i + -3, but
+    // i + -(1 + 2) will not be constant folded.
```

Comment:
Integer constant expressions of the form `+e` and `-e` are `LeafExprNodes` rather than `UnaryOperatorNodes`. The reason for this is so that they can be correctly constant folded. The only children of a `BinaryOperatorNode` that are constant folded are `LeafExprNodes` whose `Exprs` are integer constant expressions. So, for example, the expression `+ p -(1 + 2) 4` will be constant folded to `+ p 1` only if `-(1 + 2)` is a `LeafExprNode` with the `Expr` `-(1 + 2)`. If `-(1 + 2)` is a `UnaryOperatorNode`, this expression will not be constant folded. I don't think this is currently reflected in the design doc - I will update the doc so this is more clear.

The children within a `BinaryOperatorNode` will only be constant folded if the `BinaryOperatorNode` has at least two `LeafExprNode` children whose `Exprs` are integer constant expressions. For example, the `BinaryOperatorNode` + `p -(1 + 2)` will not be constant folded since it only has one integer constant child node, but `+ p -(1 + 2) 0` will be constant folded to `+ p -3`.